src/Symfony/Component/Messenger/Transport/Sender/SendersLocator.php

```diff
@@ -41,20 +42,56 @@ public function __construct(array $sendersMap, ContainerInterface $sendersLocato
      */
     public function getSenders(Envelope $envelope): iterable
     {
-        $seen = [];
+        $senderAliases = [];
 
         foreach (HandlersLocator::listTypes($envelope) as $type) {
+            $typeSenderAliases = [];
+
             foreach ($this->sendersMap[$type] ?? [] as $senderAlias) {
-                if (!\in_array($senderAlias, $seen, true)) {
-                    if (!$this->sendersLocator->has($senderAlias)) {
-                        throw new RuntimeException(sprintf('Invalid senders configuration: sender "%s" is not in the senders locator.', $senderAlias));
-                    }
-
-                    $seen[] = $senderAlias;
-                    $sender = $this->sendersLocator->get($senderAlias);
-                    yield $senderAlias => $sender;
-                }
+                $typeSenderAliases[] = $senderAlias;
+            }
+
+            if (empty($typeSenderAliases)) {
+                $typeSenderAliases = $this->getSendersFromAttributes($type);
+            }
+
+            $senderAliases = array_merge($senderAliases, $typeSenderAliases);
+        }
+
+        $senderAliases = array_unique($senderAliases);
+
+        foreach ($senderAliases as $senderAlias) {
+            if (!$this->sendersLocator->has($senderAlias)) {
+                throw new RuntimeException(sprintf('Invalid senders configuration: sender "%s" is not in the senders locator.', $senderAlias));
             }
+
+            $sender = $this->sendersLocator->get($senderAlias);
+            yield $senderAlias => $sender;
+        }
+    }
+
+    /**
+     * @return string[]
+     */
+    private function getSendersFromAttributes(string $type): array
+    {
+        if (!class_exists($type) && !interface_exists($type)) {
+            return [];
         }
+
+        try {
+            $reflectionClass = new \ReflectionClass($type);
+        } catch (\ReflectionException $e) {
+            return [];
+        }
+
+        $attributes = $reflectionClass->getAttributes(Transport::class);
+
+        return array_map(function (\ReflectionAttribute $attribute): string {
+            /** @var Transport $attributeInstance */
+            $attributeInstance = $attribute->newInstance();
+
+            return $attributeInstance->name;
+        }, $attributes);
```

Comment:
It's unfortunate to need to do this at runtime (and not compile time). But as messages are not services... or tagged in any way, there is no way for us to collect the attribute information at container build time. So 👍 for this approach.

Reply:
This is indeed unfortunate. Any clever ideas? I understand _it depends_, but is there a non-negligible performance impact to processing the attributes at runtime?